src/dataflow/analysis/flow-graph.ts

```diff
@@ -72,20 +72,26 @@ export class FlowGraph {
     return finalResult;
   }
 
-  /** Validates a single check (on the given edge). Returns true if the check passes. */
+  /** 
+   * Validates a single check (on the given edge). We define validation as
+   * every path ending at the edge must pass the check on that edge. 
+   * Returns true if the check passes.
+   */
   private validateSingleEdge(edgeToCheck: Edge): ValidationResult {
     assert(!!edgeToCheck.check, 'Edge does not have any check conditions.');
 
     const check = edgeToCheck.check;
+    const conditions = checkToList(check);
+
     const finalResult = new ValidationResult();
 
     // Check every input path into the given edge.
     // NOTE: This is very inefficient. We check every single check condition against every single edge in every single input path.
     for (const path of allInputPaths(edgeToCheck)) {
-      if (!evaluateCheckForPath(check, path)) {
-        const edgesInPath = path.edges.slice().reverse();
-        const pathString = edgesInPath.map(e => e.label).join(' -> ');
-        finalResult.failures.push(`'${check.toManifestString()}' failed for path: ${pathString}`);
+      const claimsForPath = computeTagClaimsInPath(path);
```

Comment:
Suggestion: rename claimsForPath to tagsForPath

Reply:
Done.